### PR TITLE
[wip] GRPC client

### PR DESCRIPTION
**Pull request by diocas**
Hi,

This PR is a work in progress and the idea is to start the discussion on how to provide a GRPC client for Phoenix. The Phoenix PR is here: https://github.com/owncloud/phoenix/pull/2613

## Motivation and Context
At CERN we want to test the feasibility of using GRPC between REVA and the webapp. As it was agreed with Owncloud, to accomplish this the owncloud-sdk should define a very clear interface, usable by Phoenix, and it should be modular in order to easily replace the current webdav implementation with something else. But this is not the case at the moment.

## Description
This PR creates new class representations of elements that Phoenix requires: for now I have a `File` class and I'm working on others like `FileTrashBin`, `Share`, etc. If you already have any documentation on this, please let me know. If not, let's discuss what is required. 
_These new classes have no knowledge about webdav, they are completely agnostic_, and they replace the current `SomethingInfo`. The responsibility to convert from webdav to our internal representation is inside the `Converter` functions. GRPC works in the same way.

I'm also removing completely the possibility of Phoenix to define what "webdav properties" to request, thus moving all of these from Phoenix to sdk. This is the only way we could provide a common interface for any access type! Phoenix should only know about the sdk interfaces/objects and nothing else. This also means that we need to clarify what methods are required (again, if you have some internal documentation, can I get more details about it?). I'm saying this because I see things like, for example, `mkdir`, `createFolder`, `recursiveMkdir`. Aren't these, more or less, the same thing?

Finally, this PR has a REVA/GRPC client and a `FileManagement` class that allows listing a directory. This is already usable, so you can list files in Phoenix, but nothing else. I'll work on the remaining methods once we agree on them.

Please ignore the documentation.

What do you think?
Diogo

### CODE REVIEW

**Comment by labkode** [comment]
@felixboehm @DeepDiver1975 please give feedback :)

**Comment by diocas** [comment]
For me it's perfectly fine to have separate implementations. What's missing are the interfaces.
And it was just easier to prototype both together so I could move between one and the other easily. 
Do not forget this is a prototype.

**Comment by labkode** [comment]
> Using build time tools (yarn/npm) everybody can pick the desired implementations.

I think that would be the best approach, also if you want to Graph APIs it should be done in the same way WebDAV and Grpc is done (abstracted from the developers).

@DeepDiver1975 any idea on how to hook the transport library in the build process? I imagine it requires changes in Phoenix itself

File: src/grpc/reva.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      this.clientSecret = options.auth.bearer
+      this.internalCLient = new GatewayServiceClient(this.grpcProxy, {}, {})
+      this._authenticate()
+    }
+  }
+
+  _authenticate () {
+    let authRequest = new AuthenticateRequest()
+    authRequest.setType('bearer')
+    authRequest.setClientId(this.clientId)
```
[the remark this answers is not in the document]
Comment (labkode) [reply src/grpc/reva.js]: Reva supports various auth methods, the "bearer" is just one of them.

We don't want to use the OIDC from Phoenix in every requests for many reasons and we exchange it for our own session token. This code is the token exchange part.

[the remark this answers is not in the document]
Comment (diocas) [reply src/grpc/reva.js]: This is how authentication is implemented in REVA.
ClientId and ClientSecret are the names of the variables defined in the authentication so I used the same for the internal variables. But the secret is actually the bearer.

**Comment by DeepDiver1975** [comment]
> any idea on how to hook the transport library in the build process? I imagine it requires changes in Phoenix itself

I could image a possible approach is:
- yarn remove owncloud-sdk
- yarn add owncloud-sdk-grpc

Or have both required by yarn but a build switch is choosing which to import
```js
if (env.BUILD_WITH_GRPC) {
    import sdk from 'owncloud-sdk-grpc'
} else {
    import sdk from 'owncloud-sdk'
}
```

One need to look deeper on how to do this

File: src/grpc/reva.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply src/grpc/reva.js]: understood - just saying that this is confusing ....

**Comment by DeepDiver1975** [comment]
Failing tests:
```

ERROR in ./tests/XmlParserTest.js
--

**Comment by diocas** [comment]
This is a PR to kick start the discussion only, as I mentioned before...
I did not fix the tests (in this case I would have to change the import path from `../src/xmlParser` to `../src/webdav/xmlParser`) nor finished everything. We need to define first what path to follow, as I don't want to waste time in code that is going to be thrown away.

**Comment by DeepDiver1975** [comment]
Ideas on build time decision on which transport to be used: https://stackoverflow.com/questions/28572380/conditional-build-based-on-environment-using-webpack

**Pull request closed without merging**